src/with_clause_parser.c

```diff
@@ -0,0 +1,178 @@
+/*
+ * This file and its contents are licensed under the Apache License 2.0.
+ * Please see the included NOTICE for copyright information and
+ * LICENSE-APACHE for a copy of the license.
+ */
+#include <postgres.h>
+#include <c.h>
+
+#include <commands/defrem.h>
+#include <nodes/parsenodes.h>
+#include <utils/builtins.h>
+
+#include "with_clause_parser.h"
+
+/*
+ * Filter a list of DefElem based on a namespace.
+ * This function will iterate through DefElem and output up to two lists:
+ *         within_namespace: every element within the namespace
+ *     not_within_namespace: all the other elements
+ *
+ * That is, given a with clause like:
+ *     WITH (foo.foo_para, bar.bar_param, baz.baz_param)
+ *
+ * ts_with_clause_filter(elems, "foo", in, not_in) will have
+ *        in = foo.foo_para
+ *    not_in = bar.bar_param, baz.baz_param
+ */
+void
+ts_with_clause_filter(const List *def_elems, const char *namespace, List **within_namespace,
+					  List **not_within_namespace)
+{
+	ListCell *cell;
+
+	foreach (cell, def_elems)
+	{
+		DefElem *def = (DefElem *) lfirst(cell);
+
+		if (def->defnamespace != NULL && pg_strcasecmp(def->defnamespace, namespace) == 0)
+		{
+			if (within_namespace != NULL)
+				*within_namespace = lappend(*within_namespace, def);
+		}
+		else if (not_within_namespace != NULL)
+		{
+			*not_within_namespace = lappend(*not_within_namespace, def);
+		}
+	}
+}
+
+/*
+ * Deserialize and apply the values in a WITH clause based on the on_arg table.
+ *
+ * This function will go through every element in def_elems and search for a
+ * corresponding argument in args, if one is found it will attemt to deserialize
+ * the argument, using that table elements deserialize function, then apply it
+ * to state.
+ *
+ * This is used to turn the list into a form more useful for our internal
+ * functions
+ */
+void
+ts_with_clauses_apply(const List *def_elems, ts_with_clause_on_arg on_arg,
```

Comment:
this function should return an array instead of have a callback function. The array should either be a simple datum array or something like struct {Datum, is_default}. Although I think if the defaults are well-defined there should be no reason to differentiate between the set-to-default case and the user-set-to-same-value-as-default case. So, a simple Datum array return should be sufficient.